### DISCUSSION TITLE
Tags vs meta for paused states

### DISCUSSION THREAD

**Discussion by tom-sherman** [erikras/remix-conf-2022#5] [General]
Wondering if you had any thoughts on this?

Specifically this line: https://github.com/erikras/remix-conf-2022/blob/7f489b9d98c13342000f667580c549305ad024a2/app/asyncInterpret.js#L17

Our implementation started off using tags, it was great and really ergonomic. However we had to switch to a `meta: { paused: true }` approach after we needed to support parallel paused states.

Let me see if I can articulate with an example:

```js
const fileMachine = createMachine({
  id: "file",
  type: "parallel",
  states: {
    upload: {
      initial: "idle",
      states: {
        idle: {
          tags: ["paused"],
          on: {
            INIT_UPLOAD: { target: "pending" },
          },
        },
        pending: {
          src: "uploadFile",
          onDone: "success"
        },
        success: {
          tags: ["paused"],
        },
      },
    },
    download: {
      initial: "idle",
      states: {
        idle: {
          tags: ["paused"],
          on: {
            INIT_DOWNLOAD: { target: "pending" },
          },
        },
        pending: {
          src: "downloadFile",
          onDone: "success"
        },
        success: {
          tags: ["paused"],
        },
      },
    },
  },
});
```

In this example, asyncInterpret will resolve as soon as we enter _either_ `download.success` _or_ `upload.success`. This is because tags are applied to the whole `State` and not each value. This makes a tag based approach similar to `Promise.race`

Instead if we take the above machine and use meta instead:

```js
const fileMachine = createMachine({
  id: "file",
  type: "parallel",
  states: {
    upload: {
      initial: "idle",
      states: {
        idle: {
          meta: {
            paused: true,
          },
          on: {
            INIT_UPLOAD: { target: "pending" },
          },
        },
        pending: {
          src: "uploadFile",
          onDone: "success",
        },
        success: {
          meta: {
            paused: true,
          },
        },
      },
    },
    download: {
      initial: "idle",
      states: {
        idle: {
          meta: {
            paused: true,
          },
          on: {
            INIT_DOWNLOAD: { target: "pending" },
          },
        },
        pending: {
          src: "downloadFile",
          onDone: "success",
        },
        success: {
          meta: {
            paused: true,
          },
        },
      },
    },
  },
});
```

This allows asyncInterpreter to wait for when we are in _both_ `download.success` _and_ `upload.success`. In this way it acts similar to `Promise.all`

I think the meta approach is better as waiting for all paused states is probably what you want most of the time, and you can still implement something like a `Promise.race` explicitly in the machine by transitioning to a parent state.

Any thoughts? 😄

**Comment by WangLarry** [comment]
how to waitfor the desired state with meta?

**Comment by WangLarry** [comment]
If waiting parallel state, can put tags in onDone node of parallel.

**Comment by tom-sherman** [comment]
> how to waitfor the desired state with meta?

If you declare your paused states as `meta: { paused: true }` you can enumerate the state paths, potentially with `state.configuration` to check that all state values are paused. I unfortunately don't have an OSS implementation to share yet but I should do soon.

> If waiting parallel state, can put tags in onDone node of parallel.

You can put the tag on the onDone target, but this will short circuit and not wait for other non-paused states to settle - this could cause bugs or memory leaks. Hopefully my example illustrates this, tags work essentially as `Promise.race`.

**Comment by johtso** [comment]
Couldn't you also just invert and have a `busy` tag? That would mean that if _any_ `busy` states are active it will not resolve.. may be missing something though!

**Reply by tom-sherman** [reply]
I think this would work perfectly and allow you to still use tags!

I still think using meta is more flexible though. With a paused tag you don't have any way of knowing which state is paused but this is trivial when using meta.

**Reply by Andarist** [reply]
This is using this tag in a pretty generic context - in which it doesn't matter what state is paused. I think that tags actually fit this use case perfectly.

**Reply by tom-sherman** [reply]
Not to mention the visual editor doesn't support meta and deletes the property from your machine definition 😭
